test/e2e/storage/external/external.go

```diff
@@ -0,0 +1,217 @@
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// List of testSuites to be executed for each external driver.
+var csiTestSuites = []func() testsuites.TestSuite{
+	testsuites.InitVolumesTestSuite,
+	testsuites.InitVolumeIOTestSuite,
+	testsuites.InitVolumeModeTestSuite,
+	testsuites.InitSubPathTestSuite,
+	testsuites.InitProvisioningTestSuite,
+}
+
+func init() {
+	flag.Var(testDriverParameter{}, "storage.testdriver", "name of a .yaml or .json file that defines a driver for storage testing, can be used more than once")
+}
+
+type testDriverParameter struct {
+}
+
+func (t testDriverParameter) String() string {
+	return "<.yaml or .json file>"
+}
+
+func (t testDriverParameter) Set(filename string) error {
+	driver, err := t.loadDriverDefinition(filename)
+	if err != nil {
+		return err
+	}
+
+	Describe("External Storage "+testsuites.GetDriverNameWithFeatureTags(driver), func() {
+		f := framework.NewDefaultFramework(driver.ShortName)
+
+		BeforeEach(func() {
+			// Reset config. The driver might have modified its copy
+			// in a previous test.
+			driver.GetDriverInfo().Config = testsuites.TestConfig{
+				Prefix:    driver.ShortName,
+				Framework: f,
+			}
+
+			// setupDriver
+			driver.CreateDriver()
+		})
+
+		AfterEach(func() {
+			// Cleanup driver
+			driver.CleanupDriver()
+		})
+
+		testsuites.RunTestSuite(f, driver, csiTestSuites,
+			func(patterns []testpatterns.TestPattern) []testpatterns.TestPattern {
+				return patterns
+			})
+	})
+
+	return nil
+}
+
+func (t testDriverParameter) loadDriverDefinition(filename string) (*driverDefinition, error) {
+	if filename == "" {
+		return nil, errors.New("missing file name")
+	}
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	// Some reasonable defaults follow.
+	driver := &driverDefinition{
+		DriverInfo: testsuites.DriverInfo{
+			SupportedFsType: sets.NewString(
+				"", // Default fsType
+			),
+		},
+		ClaimSize: "5Gi",
+	}
+	// TODO: strict checking of the file content once https://github.com/kubernetes/kubernetes/pull/71589
+	// or something similar is merged.
+	if err := runtime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), data, driver); err != nil {
+		return nil, errors.Wrap(err, filename)
+	}
+	return driver, nil
+}
+
+var _ testsuites.TestDriver = &driverDefinition{}
+
+// We have to implement the interface because dynamic PV may or may
+// not be supported. driverDefinition.SkipUnsupportedTest checks that
+// based on the actual driver definition.
+var _ testsuites.DynamicPVTestDriver = &driverDefinition{}
+
+// Needed only for deserialization which isn't done, therefore the
+// functions below can be empty stubs.
+var _ runtime.Object = &driverDefinition{}
+
+// DriverDefinition needs to be filled in via a .yaml or .json
+// file. It's methods then implement the TestDriver interface, using
+// nothing but the information in this struct.
+type DriverDefinition struct {
```

Comment:
Is this struct meant to be exportable?

Reply:
No, that's not necessary. I can change that.